### ISSUE TITLE
Running detekt in "type resolution" mode ignores "excludes" specified in gradle

### ISSUE THREAD

**Issue by dimsuz** [detekt/detekt#4127]
## Expected Behavior
When running with type resolution enabled (using `detektMain` or `detektDebug`, etc), Detekt should respect excludes specified with 

```groovy
subprojects {
  apply plugin: "io.gitlab.arturbosch.detekt"
  detekt { ... }

  tasks.withType(io.gitlab.arturbosch.detekt.Detekt).configureEach {
    exclude("**/generated/**")
  }
}
```

## Observed Behavior
Currently, this `exclude` option is ignored and I have files like
```
module/src/main/kotlin/data/network/entity/generated/AmountModel.kt
```
being checked by detekt and errors are reported for them.
When running without type resolution (`detekt` task), no errors are reported.

Maybe this is somehow caused by the fact that

```
tasks.withType<com....Detect> { exclude(...) }
```

while type resolution has tasks like `detektMain`, `detektDebug`, etc

## Steps to Reproduce
* Create a project with several modules, apply configuration specified above
* Create a folder `src/main/com/example/data/generated` with some file and some detektable error within
* Run detekt and observe that file isn't ignored

## Your Environment

* Version of detekt used: 1.18.1
* Version of Gradle used (if applicable): 7.0.2
* Operating System and version: Linux

**Comment by cortinico** [comment]
@BraisGabin do you recall what was the root cause of this?
I've done a quick check of the Gradle code, and it looks legit to me. The `Detekt` type of tasks are `SourceTasks` and the `exclude()` should be honored.

@dimsuz Can you provide a reproducer? Is it an Android/JVM/KMP project? Specifically I've the suspect that for the Android projects, there might be a race condition of some form.

**Comment by dimsuz** [comment]
Yep, it was an Android project with several "custom" build types and I ran the `detektDebug` task! I will try to come up with a minimal reproducer project.

**Comment by BraisGabin** [comment]
The gradle code is legit, is the cli/core that ignores it. When we use type solving we create a `List<KtFile>` and we use that list to generate the `BindingContext`. And then we run all the rules agains that `List`. We ignore the "exclude" but we also ignore the "include". I'm talking on the top of my head. I read that code time ago and it could be change.

**Comment by cortinico** [comment]
> When we use type solving we create a `List<KtFile>` and we use that list to generate the `BindingContext`

Where is this happening exactly? To me knownledge we use the `.getSource()` field from the task, that is correctly accounting for includes/excludes without us having to manually handle the `FileTree`

**Comment by BraisGabin** [comment]
I can be wrong here but I'm refering to the code in `Lifecycle.kt`:

```kotlin
        val filesToAnalyze = measure(Phase.Parsing) { parsingStrategy.invoke(settings) }
        val bindingContext = measure(Phase.Binding) { bindingProvider.invoke(filesToAnalyze) }
```

If you check here we use the same list `filesToAnalyze` to create `bindingContext`. So we can't filter files out in the gradle plugin, otherwise we would have compile errors. So, even if the error is in the gradle plugin we can't fix it that easily.

But we could create an unit test in the gradle plugin excluding a file and ensure that we aren't passing that file to the `cli`.

**Comment by eygraber** [comment]
I had a similar issue and resolved it by doing:

```
tasks.withType<Detekt>().configureEach {
  exclude {
    "foo" in it.file.absolutePath
  }
}
```

**Comment by chao2zhang** [comment]
#4151 didn't make the way to RC2 - I would appreciate if you could help test using the latest snapshot (https://detekt.github.io/detekt/snapshots.html)

**Comment by dimsuz** [comment]
Sorry for the late answer.

Still having this issue with detekt `1.21.0`.

My config:

```groovy
detekt {
  config = files(rootProject.file("detekt-config.yml"))
  ignoredBuildTypes = [ "release" ]
  ignoredFlavors = [ "<flavor>" ]
}

tasks.withType(io.gitlab.arturbosch.detekt.Detekt).configureEach {
  jvmTarget = "1.8"
  exclude("**/generated/**")
  reports {
    xml.required.set(false)
    html.required.set(false)
    txt.required.set(false)
    sarif.required.set(false)
  }
}
```
Detekt in type resolution mode:
```
lib/pathfinder-store/build/generated/sqldelight/code/PathfinderDatabase/debug/ru/kode/pathfinder/store/PathfinderDatabase.kt:23:51: Function parameter names should match the pattern: [a-z][A-Za-z0-9]* [FunctionParameterNaming]
lib/pathfinder-store/build/generated/sqldelight/code/PathfinderDatabase/debug/ru/kode/pathfinder/store/PathVariableQueries.kt:24:24: Function parameter names should match the pattern: [a-z][A-Za-z0-9]* [FunctionParameterNaming]
lib/pathfinder-store/build/generated/sqldelight/code/PathfinderDatabase/debug/ru/kode/pathfinder/store/PathVariableQueries.kt:27:5: Function parameter names should match the pattern: [a-z][A-Za-z0-9]* [FunctionParameterNaming]
```

**Comment by dimsuz** [comment]
A variation of config provided by @eygraber above helped me too:

```groovy
  exclude {
    it.file.absolutePath.contains("generated")
  }
```
(`in` didn't work for some reason)

If this is the correct way to do this, perhaps it should be mentioned in docs on https://detekt.dev.
